add init option

ranyitz opened the pull request:
## Summary
Following the feature request in #6403 this PR adds `jest --init` option.

When a user type `jest --init` in a project directory, he will be prompted with questions that will help jest to create a basic configuration, suitable for the project. It will create a `jest.config.js` file and modify `package.json` if needed.

![jest-init](https://user-images.githubusercontent.com/11733036/41262211-36b06a14-6de8-11e8-97d6-29c937547077.gif)

### What's inside?
* [prompts](https://github.com/terkelg/prompts) package added.
* `jest-config`
	* descriptions module created - reusable short explanations for configuration options.
	* `cacheDirectory`, the dynamic bit of the `defaults` module, now has it's own module so it'll be easier to mock.
* `jest-cli`
	* the `--init` options triggers the init process.

### Thoughts
* Right now we're generating the config using all of the options, maybe we can omit options like `cacheDirectory` which are less likely to be changed by the user.
* I could not write a proper e2e because I'm not sure how to fake the user answers to the prompt without mocking.

### How does the generated/modified files looks like?

`jest.config.js` 
```js
// For a detailed explanation regarding each configuration property, visit:
// https://facebook.github.io/jest/docs/en/configuration.html

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after the first failure
  // bail: false,

  // Respect "browser" field in package.json when resolving modules
  // browser: false,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/var/folders/tz/39sb86z96tj_2cqqkm1kfkh9bwd75d/T/jest_ocrnot",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: null,

  // The directory where Jest should output its coverage files
  // coverageDirectory: null,

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: null,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files usin a array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: null,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: null,

  // A set of global variables that need to be available in all test environments
  globals: {
    "ts-jest": {
      "tsConfigFile": "tsconfig.json"
    }
  },

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  moduleFileExtensions: [
    "ts",
    "tsx",
    "js"
  ],

  // A map from regular expressions to module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "always",

  // A preset that is used as a base for Jest's configuration
  // preset: null,

  // Run tests from one or more projects
  // projects: null,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: null,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: null,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // The path to a module that runs some code to configure or set up the testing framework before each test
  // setupTestFrameworkScriptFile: null,

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: "node",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  testMatch: [
    "**/__tests__/*.+(ts|tsx|js)"
  ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern Jest uses to detect test files
  // testRegex: "",

  // This option allows the use of a custom results processor
  // testResultsProcessor: null,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "about:blank",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  transform: {
    "^.+\\.(ts|tsx)$": "ts-jest"
  },

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: null,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};

```

`package.json`
```json
{
  "name": "test-jest-init",
  "version": "1.0.0",
  "main": "jest.config.js",
  "dependencies": {
    "typescript": "^2.9.1"
  },
  "scripts": {
    "test": "jest"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "description": ""
}
```

## Test plan
Verify that each file system condition (e.g. has jest config, has typescript dependency) affects the prompts, and that each response to the prompts changes the way files will e generated.

saved 1 snapshot of the whole generated file with comments, and evaluate each permutation to verify the returned value from the config.

ranyitz commented:
`prompts` currently not support node 6, [filed an issue](https://github.com/terkelg/prompts/issues/64)

SimenB commented on packages/jest-cli/src/lib/init/generate_config_file.js:
```diff
@@ -0,0 +1,94 @@
+
+  const overrideKeys: Array<string> = Object.keys(overrides);
+
+  const stringifyOption = (
```
does this have to exist inside the `generateConfigFile` closure?

ranyitz replied:
I'll move it out of there 😉

SimenB commented on packages/jest-cli/src/lib/init/generate_config_file.js:
```diff
@@ -0,0 +1,94 @@
+
+  const overrides: Object = {};
+
+  if (typescript) {
```
I'm not sure if we should add ts-jest - it forces people to install another dependency and babel 7 supports typescript..

@rickhanlonii @thymikee @orta thoughts?

thymikee replied:
Also to the problem with `ts-jest` is that it may happen that older version will be installed, e.g. right after new Jest major release. Maybe we can add another prompt?

orta replied:
I feel like this isn't that much of a problem, I've yet to have ts-jest break between jest versions (but I don't see every issue on this repo)  and it's triggered on a prompt that checks your dependencies for typescript itself, so using babel's transformer won't necessarily make sense for that kind of setup by default.

SimenB commented:
I'm really excited by this, thanks for working on it!

Code wise I think it looks good, but it would be awesome to have some docs on it (plus a changelog entry 🙂)

I'm honestly fine with this just supporting node 8 and up (as it's just something that would be run on a developer's machine), but we'd still need `prompts` to at least drop the `engine` requirement and implement a more manual "cannot be run on older than node 8" thing.

ranyitz commented:
@SimenB @thymikee Thanks for the review.

I've added docs and a changelog entry. I've also realized that it makes sense to put `jest --init` in the getting started guide, but only after the original "getting started" which instruct the user to add
```js
{
  "scripts": {
    "test": "jest"
  }
}
```

So i've added support for this case, and now we'll not ask the user that add it if it's already there.
_____

I agree that we need to decide about how we initialize jest for a typescript project, and that what we currently have is only OKish and has some problems.

Regarding prompts, opened [another issue with the update](https://github.com/terkelg/prompts/issues/66) asking to remove the `engines` field.

ranyitz commented:
**Update**, prompts now [supports node 6](https://github.com/terkelg/prompts/pull/67).

SimenB commented on yarn.lock:
```diff
@@ -7546,6 +7550,13 @@ promise@^7.1.1:
   dependencies:
     asap "~2.0.3"
 
+prompts@^0.1.8:
```
mind running `yarn` so this is updated?

rickhanlonii commented:
Agree with @thymikee, we should move this to `jest-init`

SimenB commented:
@ranyitz mind mocking out `console.log` in the tests? Quite noisy when I run them.

Asserting on the output might make sense, but if not, just stub it out 🙂

ranyitz commented:
@SimenB I agree, I'll do it later on today.
